StorageTests/TestSQLiteBookmarks.swift

```diff
@@ -77,6 +79,188 @@ class TestSQLiteBookmarks: XCTestCase {
+        ]
+
+        db.run([
+            (sql: localQuery, args: localArgs),
+            (sql: structureQuery, args: structureArgs),
+            (sql: bufferQuery, args: bufferArgs),
+            (sql: bufferStructureQuery, args: bufferStructureArgs),
+        ]).succeeded()
+
+        let expectation = self.expectation(description: #function)
+        bookmarks.getLocalBookmarksAdditions(limit: 2) >>== { results in
+            XCTAssertEqual(results.count, 2)
+            XCTAssertEqual(results.map { $0.guid }, ["bookmark_good_nonsynced", "bookmark_good_additional"])
+            expectation.fulfill()
+        }
+        waitForExpectations(timeout: 10, handler: nil)
+    }
+
+    func testApplyBufferUpdatedCompletionOp() {
+        guard let db = getBrowserDB("TSQLBtestApplyBufferUpdatedCompletionOp.db", files: self.files) else {
+            XCTFail("Unable to create browser DB.")
+            return
+        }
+
+        let localQuery =
+        "INSERT INTO \(TableBookmarksLocal) (guid, type, bmkUri, title, parentid, parentName, sync_status) " +
+        "VALUES " +
+        "(?, \(BookmarkNodeType.bookmark.rawValue), ?, ?, ?, '', 2), " +
+        "(?, \(BookmarkNodeType.bookmark.rawValue), ?, ?, ?, '', 2), " +
+        "(?, \(BookmarkNodeType.bookmark.rawValue), ?, ?, ?, '', 2), " +
+        "(?, \(BookmarkNodeType.bookmark.rawValue), ?, ?, ?, '', 2) "
+
+        let localArgs: Args = [
+            "bkm1", "http://example.org/1", "Bookmark 1", BookmarkRoots.MobileFolderGUID,
+            "bkm2", "http://example.org/2", "Bookmark 2", BookmarkRoots.MobileFolderGUID,
+            "bkm3", "http://example.org/3", "Bookmark 3", BookmarkRoots.MobileFolderGUID,
+            "bkm4", "http://example.org/4", "Bookmark 4", BookmarkRoots.MobileFolderGUID,
+        ]
+
+        let structureQuery =
+        "INSERT INTO \(TableBookmarksLocalStructure) (parent, child, idx) VALUES " +
+        "(?, ?, ?), " +
+        "(?, ?, ?), " +
+        "(?, ?, ?), " +
+        "(?, ?, ?) "
+
+        let structureArgs: Args = [
+            BookmarkRoots.MobileFolderGUID, "bkm1", 0,
+            BookmarkRoots.MobileFolderGUID, "bkm2", 1,
+            BookmarkRoots.MobileFolderGUID, "bkm3", 2,
+            BookmarkRoots.MobileFolderGUID, "bkm4", 3,
+        ]
+
+        let bufferQuery =
+        "INSERT INTO \(TableBookmarksBuffer) (guid, type, bmkUri, title, parentid, parentName, server_modified) VALUES " +
+        "(?, \(BookmarkNodeType.folder.rawValue), NULL, ?, ?, '', ?), " +
+        "(?, \(BookmarkNodeType.folder.rawValue), NULL, ?, ?, '', ?), " +
+        "(?, \(BookmarkNodeType.bookmark.rawValue), ?, ?, ?, '', ?) "
+
+        let bkmBufModified = Date.now()
+        let bufferArgs: Args = [
+            BookmarkRoots.RootGUID, "", BookmarkRoots.RootGUID, Date.now(),
+            BookmarkRoots.MobileFolderGUID, "Mobile Bookmarks", BookmarkRoots.RootGUID, Date.now(),
```

Comment:
Set these to fixed timestamps, too. After all, you want to make sure that the Mobile Bookmarks timestamp has changed after upload, and the root has not!